File: internal/parser/parser_test.go
// Copyright 2026 victorybhg
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package parser

import (
	"testing"

	"zinc/internal/lexer"
)

func parse(src string) (*Program, *Parser) {
	l := lexer.New(src)
	tokens := l.Tokenize()
	p := New(tokens)
	prog := p.Parse()
	return prog, p
}

func assertNoErrors(t *testing.T, p *Parser) {
	t.Helper()
	if len(p.Errors) > 0 {
		t.Fatalf("parse errors: %v", p.Errors)
	}
}

func TestParseFnDecl(t *testing.T) {
	prog, p := parse(`main() { }`)
	assertNoErrors(t, p)
	if len(prog.Decls) != 1 {
		t.Fatalf("expected 1 decl, got %d", len(prog.Decls))
	}
	fn, ok := prog.Decls[0].(*FnDecl)
	if !ok {
		t.Fatal("expected FnDecl")
	}
	if fn.Name != "main" {
		t.Errorf("expected name 'main', got %q", fn.Name)
	}
}

func TestParseFnWithParams(t *testing.T) {
	prog, p := parse(`Int add(Int a, Int b) { return a }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	if len(fn.Params) != 2 {
		t.Fatalf("expected 2 params, got %d", len(fn.Params))
	}
	if fn.Params[0].Name != "a" {
		t.Errorf("expected param name 'a'")
	}
	if fn.ReturnType == nil {
		t.Error("expected return type")
	}
}

func TestParseVarStmt(t *testing.T) {
	prog, p := parse(`main() { x := 42 }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	v, ok := fn.Body.Stmts[0].(*VarStmt)
	if !ok {
		t.Fatal("expected VarStmt")
	}
	if v.Name != "x" {
		t.Errorf("expected var name 'x', got %q", v.Name)
	}
}

func TestParseIfStmt(t *testing.T) {
	prog, p := parse(`main() { if x { } }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	_, ok := fn.Body.Stmts[0].(*IfStmt)
	if !ok {
		t.Fatal("expected IfStmt")
	}
}

func TestParseIfElse(t *testing.T) {
	prog, p := parse(`main() { if x { } else { } }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	stmt := fn.Body.Stmts[0].(*IfStmt)
	if stmt.ElseStmt == nil {
		t.Error("expected else branch")
	}
}

func TestParseWhile(t *testing.T) {
	prog, p := parse(`main() { while true { } }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	_, ok := fn.Body.Stmts[0].(*WhileStmt)
	if !ok {
		t.Fatal("expected WhileStmt")
	}
}

func TestParseForCStyle(t *testing.T) {
	prog, p := parse(`main() { for i := 0; i; i { } }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	f, ok := fn.Body.Stmts[0].(*ForStmt)
	if !ok {
		t.Fatal("expected ForStmt")
	}
	if f.IsRange {
		t.Error("expected C-style for, not range")
	}
}

func TestParseForIn(t *testing.T) {
	prog, p := parse(`main() { for item in items { } }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	f, ok := fn.Body.Stmts[0].(*ForStmt)
	if !ok {
		t.Fatal("expected ForStmt")
	}
	if !f.IsRange {
		t.Error("expected range for")
	}
	if f.Item != "item" {
		t.Errorf("expected item 'item', got %q", f.Item)
	}
}

func TestParseClass(t *testing.T) {
	src := `Dog {
		String name
		new(String n) { }
		pub String bark() { return name }
	}`
	prog, p := parse(src)
	assertNoErrors(t, p)
	cls, ok := prog.Decls[0].(*ClassDecl)
	if !ok {
		t.Fatal("expected ClassDecl")
	}
	if cls.Name != "Dog" {
		t.Errorf("expected class name 'Dog'")
	}
	if len(cls.Fields) != 1 {
		t.Errorf("expected 1 field, got %d", len(cls.Fields))
	}
	if cls.Ctor == nil {
		t.Error("expected constructor")
	}
	if len(cls.Methods) != 1 {
		t.Errorf("expected 1 method, got %d", len(cls.Methods))
	}
}

func TestParseInterface(t *testing.T) {
	prog, p := parse(`interface Speaker { pub String speak() }`)
	assertNoErrors(t, p)
	iface, ok := prog.Decls[0].(*InterfaceDecl)
	if !ok {
		t.Fatal("expected InterfaceDecl")
	}
	if iface.Name != "Speaker" {
		t.Errorf("expected interface name 'Speaker'")
	}
	if len(iface.Methods) != 1 {
		t.Errorf("expected 1 method sig")
	}
}

func TestParseOrHandler(t *testing.T) {
	prog, p := parse(`main() { x := risky() or { print(err) } }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	v, ok := fn.Body.Stmts[0].(*VarStmt)
	if !ok {
		t.Fatal("expected VarStmt")
	}
	if v.OrHandler == nil {
		t.Fatal("expected OrHandler on VarStmt")
	}
	if len(v.OrHandler.Body.Stmts) != 1 {
		t.Errorf("expected 1 statement in or handler body, got %d", len(v.OrHandler.Body.Stmts))
	}
}

func TestParseOrHandlerOnExprStmt(t *testing.T) {
	prog, p := parse(`main() { risky() or { print(err) } }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	es, ok := fn.Body.Stmts[0].(*ExprStmt)
	if !ok {
		t.Fatal("expected ExprStmt")
	}
	if es.OrHandler == nil {
		t.Fatal("expected OrHandler on ExprStmt")
	}
}

func TestParseOrHandlerOnAssign(t *testing.T) {
	prog, p := parse(`main() { x := 0; x = risky() or { print(err) } }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	a, ok := fn.Body.Stmts[1].(*AssignStmt)
	if !ok {
		t.Fatal("expected AssignStmt")
	}
	if a.OrHandler == nil {
		t.Fatal("expected OrHandler on AssignStmt")
	}
}

func TestParseReturnError(t *testing.T) {
	prog, p := parse(`Int risky() { return Error("oops") }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	ret, ok := fn.Body.Stmts[0].(*ReturnStmt)
	if !ok {
		t.Fatal("expected ReturnStmt")
	}
	call, ok := ret.Value.(*CallExpr)
	if !ok {
		t.Fatal("expected CallExpr as return value")
	}
	ident, ok := call.Callee.(*Ident)
	if !ok || ident.Name != "Error" {
		t.Fatal("expected Error() call")
	}
}

func TestParseImport(t *testing.T) {
	prog, p := parse(`import "fmt"`)
	assertNoErrors(t, p)
	if len(prog.Imports) != 1 {
		t.Fatalf("expected 1 import, got %d", len(prog.Imports))
	}
	if prog.Imports[0].Path != "fmt" {
		t.Errorf("expected path 'fmt', got %q", prog.Imports[0].Path)
	}
}

func TestParseImportAlias(t *testing.T) {
	prog, p := parse(`import "math/rand" as rand`)
	assertNoErrors(t, p)
	if prog.Imports[0].Alias != "rand" {
		t.Errorf("expected alias 'rand'")
	}
}

func TestParseBinaryExpr(t *testing.T) {
	prog, p := parse(`main() { x := 1 + 2 }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	v := fn.Body.Stmts[0].(*VarStmt)
	_, ok := v.Value.(*BinaryExpr)
	if !ok {
		t.Fatal("expected BinaryExpr")
	}
}

func TestParseCallExpr(t *testing.T) {
	prog, p := parse(`main() { foo(1, 2) }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	es := fn.Body.Stmts[0].(*ExprStmt)
	call, ok := es.Expr.(*CallExpr)
	if !ok {
		t.Fatal("expected CallExpr")
	}
	if len(call.Args) != 2 {
		t.Errorf("expected 2 args, got %d", len(call.Args))
	}
}

func TestParseListLit(t *testing.T) {
	prog, p := parse(`main() { x := [1, 2, 3] }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	v := fn.Body.Stmts[0].(*VarStmt)
	lit, ok := v.Value.(*ListLit)
	if !ok {
		t.Fatal("expected ListLit")
	}
	if len(lit.Elements) != 3 {
		t.Errorf("expected 3 elements, got %d", len(lit.Elements))
	}
}

func TestParseMapLit(t *testing.T) {
	prog, p := parse(`main() { m := {"key": 1} }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	v := fn.Body.Stmts[0].(*VarStmt)
	_, ok := v.Value.(*MapLit)
	if !ok {
		t.Fatal("expected MapLit")
	}
}

func TestParseConstructorNew(t *testing.T) {
	prog, p := parse(`main() { d := Dog("rex") }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	v := fn.Body.Stmts[0].(*VarStmt)
	_, ok := v.Value.(*CallExpr)
	if !ok {
		t.Fatal("expected CallExpr for Dog()")
	}
}

func TestParsePubFn(t *testing.T) {
	prog, p := parse(`pub String greet() { return "hi" }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	if !fn.IsPub {
		t.Error("expected IsPub = true")
	}
}

func TestParsePackageDecl(t *testing.T) {
	prog, p := parse(`package "myapp/utils"
Int add(Int a, Int b) { return a }`)
	assertNoErrors(t, p)
	if prog.Package == nil {
		t.Fatal("expected Package to be non-nil")
	}
	if prog.Package.Path != "myapp/utils" {
		t.Errorf("expected Path 'myapp/utils', got %q", prog.Package.Path)
	}
	if len(prog.Decls) != 1 {
		t.Errorf("expected 1 decl, got %d", len(prog.Decls))
	}
}

func TestParseNoPackageDecl(t *testing.T) {
	prog, p := parse(`main() { }`)
	assertNoErrors(t, p)
	if prog.Package != nil {
		t.Error("expected Package to be nil when not declared")
	}
}

func TestParsePackageDeclWithImports(t *testing.T) {
	prog, p := parse(`package "myapp/models"
import "fmt"
Dog { String name }`)
	assertNoErrors(t, p)
	if prog.Package == nil {
		t.Fatal("expected Package to be non-nil")
	}
	if prog.Package.Path != "myapp/models" {
		t.Errorf("expected path 'myapp/models', got %q", prog.Package.Path)
	}
	if len(prog.Imports) != 1 {
		t.Errorf("expected 1 import, got %d", len(prog.Imports))
	}
	if len(prog.Decls) != 1 {
		t.Errorf("expected 1 decl, got %d", len(prog.Decls))
	}
}

// --- Default parameter values ------------------------------------------------

func TestParseParamWithDefault(t *testing.T) {
	prog, p := parse(`greet(String name, String greeting = "Hello") {}`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	if len(fn.Params) != 2 {
		t.Fatalf("expected 2 params, got %d", len(fn.Params))
	}
	if fn.Params[0].Default != nil {
		t.Error("expected no default for first param")
	}
	if fn.Params[1].Default == nil {
		t.Fatal("expected default for second param")
	}
	lit, ok := fn.Params[1].Default.(*StringLit)
	if !ok {
		t.Fatal("expected StringLit default")
	}
	if lit.Value != "Hello" {
		t.Errorf("expected default %q, got %q", "Hello", lit.Value)
	}
}

func TestParseCtorParamWithDefault(t *testing.T) {
	prog, p := parse(`Dog {
		String name
		Int age
		new(String name, Int age = 0) {}
	}`)
	assertNoErrors(t, p)
	cls := prog.Decls[0].(*ClassDecl)
	if cls.Ctor == nil {
		t.Fatal("expected ctor")
	}
	if len(cls.Ctor.Params) != 2 {
		t.Fatalf("expected 2 ctor params, got %d", len(cls.Ctor.Params))
	}
	if cls.Ctor.Params[0].Default != nil {
		t.Error("expected no default for first ctor param")
	}
	if cls.Ctor.Params[1].Default == nil {
		t.Fatal("expected default for second ctor param")
	}
	lit, ok := cls.Ctor.Params[1].Default.(*IntLit)
	if !ok {
		t.Fatalf("expected IntLit default, got %T", cls.Ctor.Params[1].Default)
	}
	if lit.Value != "0" {
		t.Errorf("expected default 0, got %q", lit.Value)
	}
}

// --- Named arguments at call sites -------------------------------------------

func TestParseNamedArgs(t *testing.T) {
	prog, p := parse(`main() { foo(x: 5, y: 10) }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	call := fn.Body.Stmts[0].(*ExprStmt).Expr.(*CallExpr)
	if len(call.Args) != 0 {
		t.Errorf("expected 0 positional args, got %d", len(call.Args))
	}
	if len(call.NamedArgs) != 2 {
		t.Fatalf("expected 2 named args, got %d", len(call.NamedArgs))
	}
	if call.NamedArgs[0].Name != "x" {
		t.Errorf("expected named arg x, got %q", call.NamedArgs[0].Name)
	}
	if call.NamedArgs[1].Name != "y" {
		t.Errorf("expected named arg y, got %q", call.NamedArgs[1].Name)
	}
}

func TestParseMixedPositionalAndNamedArgs(t *testing.T) {
	prog, p := parse(`main() { greet("Bob", greeting: "Hi") }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	call := fn.Body.Stmts[0].(*ExprStmt).Expr.(*CallExpr)
	if len(call.Args) != 1 {
		t.Fatalf("expected 1 positional arg, got %d", len(call.Args))
	}
	if len(call.NamedArgs) != 1 {
		t.Fatalf("expected 1 named arg, got %d", len(call.NamedArgs))
	}
	if call.NamedArgs[0].Name != "greeting" {
		t.Errorf("expected named arg greeting, got %q", call.NamedArgs[0].Name)
	}
}

func TestParseWithStmtSingle(t *testing.T) {
	prog, p := parse(`main() { with (f := openFile("x")) { } }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	ws, ok := fn.Body.Stmts[0].(*WithStmt)
	if !ok {
		t.Fatal("expected WithStmt")
	}
	if len(ws.Resources) != 1 {
		t.Fatalf("expected 1 resource, got %d", len(ws.Resources))
	}
	if ws.Resources[0].Name != "f" {
		t.Errorf("expected resource name 'f', got %q", ws.Resources[0].Name)
	}
	if _, ok := ws.Resources[0].Value.(*CallExpr); !ok {
		t.Error("expected resource value to be CallExpr")
	}
}

func TestParseWithStmtMultiple(t *testing.T) {
	prog, p := parse(`main() { with (a := foo(), b := bar()) { } }`)
	assertNoErrors(t, p)
	fn := prog.Decls[0].(*FnDecl)
	ws, ok := fn.Body.Stmts[0].(*WithStmt)
	if !ok {
		t.Fatal("expected WithStmt")
	}
	if len(ws.Resources) != 2 {
		t.Fatalf("expected 2 resources, got %d", len(ws.Resources))
	}
	if ws.Resources[0].Name != "a" {
		t.Errorf("expected resource[0] name 'a', got %q", ws.Resources[0].Name)
	}
	if ws.Resources[1].Name != "b" {
		t.Errorf("expected resource[1] name 'b', got %q", ws.Resources[1].Name)
	}
}

File: compiler/test/zinc/compiler/ParserTest.java
// Copyright 2026 victorybhg
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0

package zinc.compiler;

import static zinc.compiler.Ast.*;

public class ParserTest {

    static int passed = 0;
    static int failed = 0;

    public static void main(String[] args) {
        testHelloWorld();
        testFunction();
        testClassDecl();
        testInterface();
        testDataClass();
        testEnum();
        testVarDecl();
        testIfElse();
        testForIn();
        testWhile();
        testMatch();
        testOrHandler();
        testLambda();
        testMultiParamLambda();
        testImports();
        testBinaryExpr();
        testMethodCall();
        testNewExpr();
        testListLit();
        testMapLit();
        testStringInterp();
        testSpawn();
        testParallelFor();
        testConcurrent();
        testSealedClass();
        testAnnotation();

        System.out.println("\nResults: " + passed + " passed, " + failed + " failed");
        if (failed > 0) System.exit(1);
    }

    static Program parse(String src) {
        var tokens = new Lexer(src).tokenize();
        var parser = new Parser(tokens);
        var prog = parser.parse();
        if (!parser.errors().isEmpty()) {
            System.out.println("  Parse errors: " + parser.errors());
        }
        return prog;
    }

    // --- Tests ---------------------------------------------------------------

    static void testHelloWorld() {
        var prog = parse("print(\"Hello, World!\")");
        expect("hello: 1 stmt", prog.stmts().size(), 1);
        expect("hello: is ExprStmt", prog.stmts().getFirst() instanceof ExprStmt, true);
    }

    static void testFunction() {
        var prog = parse("fn add(int a, int b): int { return a + b }");
        expect("fn: 1 decl", prog.decls().size(), 1);
        var fn = (FnDecl) prog.decls().getFirst();
        expect("fn: name", fn.name(), "add");
        expect("fn: 2 params", fn.params().size(), 2);
        expect("fn: return type", ((SimpleType) fn.returnType()).name(), "int");
        expect("fn: body has 1 stmt", fn.body().stmts().size(), 1);
    }

    static void testClassDecl() {
        var prog = parse("""
            class Dog : Animal {
                init String name
                var int age = 0

                init(String name) {
                    this.name = name
                }

                pub fn bark(): String {
                    return "Woof!"
                }
            }
            """);
        expect("cls: 1 decl", prog.decls().size(), 1);
        var cls = (ClassDecl) prog.decls().getFirst();
        expect("cls: name", cls.name(), "Dog");
        expect("cls: parent", cls.parents().getFirst(), "Animal");
        expect("cls: 2 fields", cls.fields().size(), 2);
        expect("cls: 1 ctor", cls.ctors().size(), 1);
        expect("cls: 1 method", cls.methods().size(), 1);
        expect("cls: method name", cls.methods().getFirst().name(), "bark");
    }

    static void testInterface() {
        var prog = parse("""
            interface Speaker {
                fn speak(): String
                fn volume(): int
            }
            """);
        var iface = (InterfaceDecl) prog.decls().getFirst();
        expect("iface: name", iface.name(), "Speaker");
        expect("iface: 2 methods", iface.methods().size(), 2);
        expect("iface: method1", iface.methods().getFirst().name(), "speak");
    }

    static void testDataClass() {
        var prog = parse("data Point(int x, int y)");
        var dc = (DataClassDecl) prog.decls().getFirst();
        expect("data: name", dc.name(), "Point");
        expect("data: 2 params", dc.params().size(), 2);
    }

    static void testEnum() {
        var prog = parse("enum Color { Red, Green, Blue }");
        var en = (EnumDecl) prog.decls().getFirst();
        expect("enum: name", en.name(), "Color");
        expect("enum: 3 variants", en.variants().size(), 3);
    }

    static void testVarDecl() {
        var prog = parse("var x = 42");
        var stmt = (VarStmt) prog.stmts().getFirst();
        expect("var: name", stmt.name(), "x");
        expect("var: value is IntLit", stmt.value() instanceof IntLit, true);
    }

    static void testIfElse() {
        var prog = parse("if x > 0 { print(\"pos\") } else { print(\"neg\") }");
        var stmt = (IfStmt) prog.stmts().getFirst();
        expect("if: has then", stmt.then().stmts().size(), 1);
        expect("if: has else", stmt.elseStmt() != null, true);
    }

    static void testForIn() {
        var prog = parse("for item in items { print(item) }");
        var stmt = (ForStmt) prog.stmts().getFirst();
        expect("for: is range", stmt.isRange(), true);
        expect("for: item", stmt.item(), "item");
    }

    static void testWhile() {
        var prog = parse("while running { process() }");
        var stmt = (WhileStmt) prog.stmts().getFirst();
        expect("while: has cond", stmt.cond() instanceof Ident, true);
        expect("while: has body", stmt.body().stmts().size(), 1);
    }

    static void testMatch() {
        var prog = parse("""
            match x {
                case 1 { print("one") }
                case 2 { print("two") }
                case _ { print("other") }
            }
            """);
        var stmt = (MatchStmt) prog.stmts().getFirst();
        expect("match: 3 cases", stmt.cases().size(), 3);
    }

    static void testOrHandler() {
        var prog = parse("var x = risky() or { \"default\" }");
        var stmt = (VarStmt) prog.stmts().getFirst();
        expect("or: has handler", stmt.orHandler() != null, true);
        expect("or: handler has body", stmt.orHandler().body() != null, true);
    }

    static void testLambda() {
        var prog = parse("var f = x -> x * 2");
        var stmt = (VarStmt) prog.stmts().getFirst();
        expect("lambda: value is lambda", stmt.value() instanceof LambdaExpr, true);
        var lam = (LambdaExpr) stmt.value();
        expect("lambda: 1 param", lam.params().size(), 1);
    }

    static void testMultiParamLambda() {
        var prog = parse("var f = (a, b) -> a + b");
        var stmt = (VarStmt) prog.stmts().getFirst();
        var lam = (LambdaExpr) stmt.value();
        expect("mlambda: 2 params", lam.params().size(), 2);
    }

    static void testImports() {
        var prog = parse("""
            import java.util.List
            import java.nio.file.Files
            print("done")
            """);
        expect("import: 2 imports", prog.imports().size(), 2);
        expect("import: path", prog.imports().getFirst().path(), "java.util.List");
    }

    static void testBinaryExpr() {
        var prog = parse("var x = 1 + 2 * 3");
        var stmt = (VarStmt) prog.stmts().getFirst();
        // Should be 1 + (2 * 3) due to precedence
        var bin = (BinaryExpr) stmt.value();
        expect("binop: top is +", bin.op(), "+");
        expect("binop: right is *", ((BinaryExpr) bin.right()).op(), "*");
    }

    static void testMethodCall() {
        var prog = parse("list.add(42)");
        var stmt = (ExprStmt) prog.stmts().getFirst();
        var call = (CallExpr) stmt.expr();
        var sel = (SelectorExpr) call.callee();
        expect("call: object", ((Ident) sel.object()).name(), "list");
        expect("call: field", sel.field(), "add");
        expect("call: 1 arg", call.args().size(), 1);
    }

    static void testNewExpr() {
        var prog = parse("var x = new ArrayList(10)");
        var stmt = (VarStmt) prog.stmts().getFirst();
        var call = (CallExpr) stmt.value();
        expect("new: isNew", call.isNew(), true);
        expect("new: callee", ((Ident) call.callee()).name(), "ArrayList");
    }

    static void testListLit() {
        var prog = parse("var items = [1, 2, 3]");
        var stmt = (VarStmt) prog.stmts().getFirst();
        var list = (ListLit) stmt.value();
        expect("list: 3 elements", list.elements().size(), 3);
    }

    static void testMapLit() {
        var prog = parse("var m = {\"a\": 1, \"b\": 2}");
        var stmt = (VarStmt) prog.stmts().getFirst();
        var map = (MapLit) stmt.value();
        expect("map: 2 keys", map.keys().size(), 2);
    }

    static void testStringInterp() {
        var prog = parse("print(\"Hello, {name}!\")");
        var stmt = (ExprStmt) prog.stmts().getFirst();
        var call = (CallExpr) stmt.expr();
        var interp = (StringInterpLit) call.args().getFirst();
        expect("interp: 3 parts", interp.parts().size(), 3);
    }

    static void testSpawn() {
        var prog = parse("""
            var t = spawn {
                print("working")
            } or {
                print("failed")
            }
            """);
        var stmt = (VarStmt) prog.stmts().getFirst();
        var spawn = (SpawnExpr) stmt.value();
        expect("spawn: has body", spawn.body().stmts().size(), 1);
        expect("spawn: has or", spawn.orHandler() != null, true);
    }

    static void testParallelFor() {
        var prog = parse("parallel(max: 4) for item in items { process(item) }");
        var stmt = (ParallelForStmt) prog.stmts().getFirst();
        expect("pfor: item", stmt.item(), "item");
        expect("pfor: max", stmt.max(), 4);
    }

    static void testConcurrent() {
        var prog = parse("""
            concurrent {
                fetchUser(id)
                fetchOrders(id)
            }
            """);
        var stmt = (ConcurrentStmt) prog.stmts().getFirst();
        expect("concurrent: 2 tasks", stmt.tasks().size(), 2);
    }

    static void testSealedClass() {
        var prog = parse("""
            sealed class Shape {
                data Circle(double radius)
                data Rect(double w, double h)
            }
            """);
        // Sealed class parsed as class with nested data decls
        expect("sealed: 1 decl", prog.decls().size(), 1);
    }

    static void testAnnotation() {
        var prog = parse("""
            @Controller
            class Foo {
                @Get("/hello")
                pub fn hello(): String {
                    return "hello"
                }
            }
            """);
        var cls = (ClassDecl) prog.decls().getFirst();
        expect("ann: class has 1 annotation", cls.annotations().size(), 1);
        expect("ann: class annotation name", cls.annotations().getFirst().name(), "Controller");
        expect("ann: method has annotation", cls.methods().getFirst().annotations().size(), 1);
    }

    // --- Helpers -------------------------------------------------------------

    static void expect(String name, Object actual, Object expected) {
        if (expected.equals(actual)) {
            passed++;
        } else {
            failed++;
            System.out.println("FAIL: " + name + " — expected " + expected + ", got " + actual);
        }
    }
}
